rvgo/cmd/run.go

```diff
@@ -0,0 +1,231 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/urfave/cli/v2"
+
+	"github.com/pkg/profile"
+
+	cannon "github.com/ethereum-optimism/optimism/cannon/cmd"
+	preimage "github.com/ethereum-optimism/optimism/op-preimage"
+	"github.com/ethereum-optimism/optimism/op-service/jsonutil"
+
+	"github.com/ethereum-optimism/asterisc/rvgo/fast"
+)
+
+type Proof struct {
+	Step uint64 `json:"step"`
+
+	Pre  common.Hash `json:"pre"`
+	Post common.Hash `json:"post"`
+
+	StateData hexutil.Bytes `json:"state-data"`
+	ProofData hexutil.Bytes `json:"proof-data"`
+
+	OracleKey    hexutil.Bytes `json:"oracle-key,omitempty"`
+	OracleValue  hexutil.Bytes `json:"oracle-value,omitempty"`
+	OracleOffset uint64        `json:"oracle-offset,omitempty"`
+}
+
+type StepFn func(proof bool) (*fast.StepWitness, error)
+
+func Guard(proc *os.ProcessState, fn StepFn) StepFn {
+	return func(proof bool) (*fast.StepWitness, error) {
+		wit, err := fn(proof)
+		if err != nil {
+			if proc.Exited() {
+				return nil, fmt.Errorf("pre-image server exited with code %d, resulting in err %w", proc.ExitCode(), err)
+			} else {
+				return nil, err
+			}
+		}
+		return wit, nil
+	}
+}
+
+var _ fast.PreimageOracle = (*ProcessPreimageOracle)(nil)
+
+var OutFilePerm = os.FileMode(0o755)
+
+func Run(ctx *cli.Context) error {
+	if ctx.Bool(cannon.RunPProfCPU.Name) {
+		defer profile.Start(profile.NoShutdownHook, profile.ProfilePath("."), profile.CPUProfile).Stop()
+	}
+
+	state, err := jsonutil.LoadJSON[fast.VMState](ctx.Path(cannon.RunInputFlag.Name))
+	if err != nil {
+		return err
+	}
+
+	l := Logger(os.Stderr, log.LvlInfo)
+	outLog := &LoggingWriter{Name: "program std-out", Log: l}
+	errLog := &LoggingWriter{Name: "program std-err", Log: l}
+
+	stopAtPreimageType := ctx.String(cannon.RunStopAtPreimageTypeFlag.Name)
+	if stopAtPreimageType != "" && stopAtPreimageType != "any" && stopAtPreimageType != "local" && stopAtPreimageType != "global" {
+		return fmt.Errorf("invalid preimage type %q, must be either 'any', 'local' or 'global'", stopAtPreimageType)
+	}
+	stopAtPreimageLargerThan := ctx.Int(cannon.RunStopAtPreimageLargerThanFlag.Name)
```

Comment:
These options are only used to help us test `op-challenger` so not strictly necessary in asterisc but for the record, they changed a bit in https://github.com/ethereum-optimism/optimism/pull/9406 so we can find the additional preimage types added for 4844 support.